File: python/promptresponse/expressions.py
"""APR's optional CEL expression profile.

The stored APR value remains a string. This module performs the format-specific
binding to CEL and never treats an expression failure as a document failure.
"""

from datetime import datetime, timezone
from typing import Mapping, Optional

from celpy import Environment, celtypes

from .models import AprDocument, Prompt

PROFILE = "core+expressions"
COMPUTED_SOURCE = "computed"


def _cel_type(expected: Optional[str]):
    return {
        "number": celtypes.DoubleType,
        "currency": celtypes.DoubleType,
        "range": celtypes.DoubleType,
        "boolean": celtypes.BoolType,
        "date": celtypes.TimestampType,
        "time": celtypes.TimestampType,
        "datetime": celtypes.TimestampType,
        "multichoice": celtypes.ListType,
    }.get((expected or "").lower(), celtypes.StringType)


def _bind(response: str, expected: Optional[str]):
    kind = (expected or "").lower()
    value = response or ""
    if kind in {"number", "currency", "range"}:
        if not value.strip():
            return None
        try:
            return float(value.strip())
        except ValueError:
            return None
    if kind == "boolean":
        normalized = value.strip().lower()
        if normalized in {"true", "yes", "y", "1", "on", "x", "checked"}:
            return True
        if normalized in {"false", "no", "n", "0", "off", "unchecked"}:
            return False
        return None
    if kind in {"date", "time", "datetime"}:
        if not value.strip():
            return None
        try:
            source = value.strip()
            if kind == "date":
                source += "T00:00:00+00:00"
            elif kind == "time":
                source = "1970-01-01T" + source + "+00:00"
            elif source.endswith("Z"):
                source = source[:-1] + "+00:00"
            return datetime.fromisoformat(source).astimezone(timezone.utc)
        except ValueError:
            return None
    if kind == "multichoice":
        return [part.strip() for part in (value.split("\n") if "\n" in value else value.split(",")) if part.strip()]
    return value


def _stored(value) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, float):
        return format(value, ".15g")
    if isinstance(value, datetime):
        return value.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    if isinstance(value, (list, tuple)):
        return "\n".join(str(item) for item in value)
    return "" if value is None else str(value)


class ExpressionContext:
    def __init__(self, document: AprDocument, today: Optional[str] = None, ctx: Optional[Mapping[str, str]] = None):
        self.document = document
        self.prompts = {prompt.id: prompt for prompt in document.all_prompts() if prompt.id}
        self.bindings = {}
        self.types = {}
        for prompt_id, prompt in self.prompts.items():
            expected = prompt.hints.expected_data_type
            self.types[prompt_id] = _cel_type(expected)
            bound = _bind(prompt.response, expected)
            if bound is not None:
                self.bindings[prompt_id] = bound
        self.types["_today"] = celtypes.TimestampType
        parsed_today = _bind(today or datetime.now(timezone.utc).isoformat(), "datetime")
        if parsed_today is not None:
            self.bindings["_today"] = parsed_today
        self.types["ctx"] = celtypes.MapType
        self.bindings["ctx"] = dict(ctx or {})

    def evaluate(self, prompt: Prompt, expression: str):
        annotations = dict(self.types)
        annotations["_this"] = _cel_type(prompt.hints.expected_data_type)
        current = _bind(prompt.response, prompt.hints.expected_data_type)
        bindings = dict(self.bindings)
        if current is not None:
            bindings["_this"] = current
        try:
            environment = Environment(annotations=annotations)
            return environment.program(environment.compile(expression)).evaluate(bindings)
        except Exception:
            return None


def build_expression_context(document: AprDocument, today: Optional[str] = None, ctx: Optional[Mapping[str, str]] = None) -> ExpressionContext:
    return ExpressionContext(document, today, ctx)


def compute_value(prompt: Prompt, context: ExpressionContext) -> Optional[str]:
    expression = prompt.hints.expr_value
    if not expression or not expression.strip():
        return None
    result = context.evaluate(prompt, expression)
    return None if result is None else _stored(result)


def condition(prompt: Prompt, expression: Optional[str], context: ExpressionContext) -> bool:
    return bool(context.evaluate(prompt, expression)) if expression and expression.strip() else False


def validation_message(prompt: Prompt, context: ExpressionContext) -> Optional[str]:
    expression = prompt.hints.expr_validation
    if not expression or not expression.strip():
        return None
    result = context.evaluate(prompt, expression)
    message = None if result is None else _stored(result)
    return message or None


def recompute_computed_values(document: AprDocument, today: Optional[str] = None, ctx: Optional[Mapping[str, str]] = None) -> bool:
    changed = False
    for _ in range(5):
        context = build_expression_context(document, today, ctx)
        changed_this_pass = False
        for prompt in document.all_prompts():
            if not prompt.hints.expr_value:
                continue
            authored = bool(prompt.response) and prompt.response_metadata.source != COMPUTED_SOURCE
            if authored:
                continue
            computed = compute_value(prompt, context)
            if computed is not None and computed != prompt.response:
                prompt.response = computed
                prompt.response_metadata.source = COMPUTED_SOURCE
                changed_this_pass = changed = True
        if not changed_this_pass:
            break
    return changed

File: typescript/src/expressions.ts
/** APR's optional CEL expression binding. Expressions are advisory and pure. */
import { Environment, parse } from "@marcbachmann/cel-js";
import type { AprDocument, Prompt, Section } from "./model.js";

type ContextValues = Record<string, string>;

// Specification: "Provide the CEL standard library and standard macros, and no
// extension library or custom function" (the CEL strings extension is
// explicitly not required). cel-js, unlike celpy, bakes the strings and bytes
// extension member functions into every environment with no option to
// disable them, so an expression using one would evaluate here and fail
// identically in Python and .NET -- a cross-SDK divergence a conformance
// suite exists to catch. Names are exactly the cel-js built-ins absent from
// celpy's base registry (functions.js's functionOverload calls, diffed
// against celpy's activation dump).
const EXTENSION_FUNCTIONS = new Set([
  "lowerAscii", "upperAscii", "trim", "indexOf", "lastIndexOf", "substring", "split", "join",
  "json", "hex", "base64", "at",
]);
function usesExtensionFunction(node: unknown): boolean {
  if (Array.isArray(node)) return node.some(usesExtensionFunction);
  if (!node || typeof node !== "object") return false;
  const { op, args } = node as { op?: string; args?: unknown };
  const name = Array.isArray(args) ? args[0] : undefined;
  if ((op === "call" || op === "rcall") && typeof name === "string" && EXTENSION_FUNCTIONS.has(name)) return true;
  return usesExtensionFunction(args);
}

function prompts(sections: Section[]): Prompt[] {
  return sections.flatMap(section => [...section.prompts, ...prompts(section.sections)]);
}
const TIMESTAMP = "google.protobuf.Timestamp";
function typeFor(expected?: string): "double" | "bool" | typeof TIMESTAMP | "list<string>" | "string" {
  switch ((expected ?? "").toLowerCase()) {
    case "number": case "currency": case "range": return "double";
    case "boolean": return "bool";
    case "date": case "time": case "datetime": return TIMESTAMP;
    case "multichoice": return "list<string>";
    default: return "string";
  }
}
function bind(value: string, expected?: string): unknown | undefined {
  const kind = (expected ?? "").toLowerCase();
  if (["number", "currency", "range"].includes(kind)) {
    if (!value.trim() || !Number.isFinite(Number(value.trim()))) return undefined;
    return Number(value.trim());
  }
  if (kind === "boolean") {
    const v = value.trim().toLowerCase();
    if (["true", "yes", "y", "1", "on", "x", "checked"].includes(v)) return true;
    if (["false", "no", "n", "0", "off", "unchecked"].includes(v)) return false;
    return undefined;
  }
  if (["date", "time", "datetime"].includes(kind)) {
    if (!value.trim()) return undefined;
    const source = kind === "date" ? `${value}T00:00:00Z` : kind === "time" ? `1970-01-01T${value}Z` : value;
    const timestamp = new Date(source);
    return Number.isNaN(timestamp.valueOf()) ? undefined : timestamp;
  }
  if (kind === "multichoice") return (value.includes("\n") ? value.split("\n") : value.split(",")).map(x => x.trim()).filter(Boolean);
  return value;
}
function stored(value: unknown): string {
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") return String(value);
  if (value instanceof Date) return value.toISOString().replace(/\.\d{3}Z$/, "Z");
  if (Array.isArray(value)) return value.map(String).join("\n");
  return value == null ? "" : String(value);
}
// The standard library's string(timestamp), which cel-js lacks: RFC 3339 in UTC,
// with fractional seconds only where the instant has them.
function timestampString(value: Date): string {
  return value.toISOString().replace(/\.(\d{3})Z$/, (_, fraction: string) => fraction === "000" ? "Z" : `.${fraction.replace(/0+$/, "")}Z`);
}

const RESERVED_NAMES = new Set(["_now", "_today", "_id", "_this", "ctx"]);

export class ExpressionContext {
  private readonly fields: Map<string, Prompt>;
  private readonly bindings: Record<string, unknown> = {};
  // _now and _today are supplied by the caller and never read from the host
  // clock: defaulting them to "now" made every expression using them
  // non-deterministic, and different from the other implementations.
  // _now is the instant the caller supplied, a separate input from _today.
  constructor(document: AprDocument, today?: string, ctx: ContextValues = {}, now?: string) {
    this.fields = new Map(prompts(document.sections).filter(prompt => prompt.id).map(prompt => [prompt.id, prompt]));
    for (const prompt of this.fields.values()) {
      const value = bind(prompt.response, prompt.hints.expectedDataType);
      if (value !== undefined) this.bindings[prompt.id] = value;
    }
    if (now) {
      const instant = bind(now, "datetime");
      if (instant !== undefined) this.bindings._now = instant;
    }
    // _today is the date as YYYY-MM-DD, a string rather than a timestamp.
    if (today) this.bindings._today = today.slice(0, 10);
    this.bindings.ctx = ctx;
  }
  evaluate(prompt: Prompt, expression: string): unknown | undefined {
    try {
      if (usesExtensionFunction(parse(expression).ast)) return undefined;
      const environment = new Environment({ unlistedVariablesAreDyn: false });
      // The activation's reserved names (specification: "that prompt's bound
      // type... where the id is a valid CEL identifier and not reserved") take
      // precedence over a same-named prompt field. cel-js throws on a second
      // registerVariable call for the same name rather than letting the later
      // one win, so a document with a field literally named "ctx" would
      // otherwise fail every expression in it via the catch below.
      for (const field of this.fields.values()) if (!RESERVED_NAMES.has(field.id)) environment.registerVariable(field.id, typeFor(field.hints.expectedDataType));
      environment
        .registerVariable("_today", "string")
        .registerVariable("_now", TIMESTAMP)
        .registerVariable("_id", "string")
        .registerVariable("ctx", "map<string, string>")
        .registerVariable("_this", typeFor(prompt.hints.expectedDataType))
        .registerFunction(`string(${TIMESTAMP}): string`, timestampString);
      const bindings = { ...this.bindings };
      const current = bind(prompt.response, prompt.hints.expectedDataType);
      if (current !== undefined) bindings._this = current;
      bindings._id = prompt.id;
      return environment.evaluate(expression, bindings);
    } catch { return undefined; }
  }
}

export function buildExpressionContext(document: AprDocument, today?: string, ctx?: ContextValues, now?: string): ExpressionContext { return new ExpressionContext(document, today, ctx, now); }
export function computeValue(prompt: Prompt, context: ExpressionContext): string | undefined {
  const expression = prompt.hints.exprValue;
  if (!expression?.trim()) return undefined;
  const value = context.evaluate(prompt, expression);
  return value === undefined ? undefined : stored(value);
}
export function condition(prompt: Prompt, expression: string | undefined, context: ExpressionContext): boolean { return expression?.trim() ? context.evaluate(prompt, expression) === true : false; }
export function validationMessage(prompt: Prompt, context: ExpressionContext): string | undefined {
  const expression = prompt.hints.exprValidation;
  if (!expression?.trim()) return undefined;
  const value = context.evaluate(prompt, expression);
  // exprValidation is typed "string" (specification's expression profile): a
  // result of any other CEL type is the same failure as a compile error, not
  // a value to stringify -- 2 + 2 is not almost a validation message.
  if (typeof value !== "string") return undefined;
  return value || undefined;
}
export function recomputeComputedValues(document: AprDocument, today?: string, ctx?: ContextValues, now?: string): boolean {
  let changed = false;
  for (let pass = 0; pass < 5; pass++) {
    const context = buildExpressionContext(document, today, ctx, now); let changedThisPass = false;
    for (const prompt of prompts(document.sections)) {
      if (!prompt.hints.exprValue) continue;
      // Every non-empty response in the document as it was read is authored, whatever
      // produced it, and APR-EXPR-001 says an expression must not rewrite one. Only what
      // this session computed may be recomputed.
      if (prompt.response && !prompt.computedInThisSession) continue;
      const value = computeValue(prompt, context);
      if (value !== undefined && value !== prompt.response) { prompt.response = value; prompt.computedInThisSession = true; changed = changedThisPass = true; }
    }
    if (!changedThisPass) break;
  }
  return changed;
}
